File: packages/types/src/errors/resource-errors.ts
/**
 * 资源相关错误类型定义
 * 定义资源未找到相关的错误类型
 *
 * 注意：所有资源未找到错误默认为严重错误（error级别）
 * 如果需要记录警告但不中断执行，请使用 ContextualLogger.resourceNotFoundWarning()
 */

import { NotFoundError, ErrorSeverity } from "./base.js";

/**
 * Workflow not found error type
 *
 * Specialized for workflow not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class WorkflowNotFoundError extends NotFoundError {
  constructor(
    message: string,
    workflowId: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "Workflow", workflowId, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Node not found error type
 *
 * Specialized for node not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class NodeNotFoundError extends NotFoundError {
  constructor(
    message: string,
    nodeId: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "Node", nodeId, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Tool not found error type
 *
 * Specialized for tool not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class ToolNotFoundError extends NotFoundError {
  constructor(
    message: string,
    toolId: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "Tool", toolId, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Script not found error type
 *
 * Specialized for script not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class ScriptNotFoundError extends NotFoundError {
  constructor(
    message: string,
    scriptName: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "Script", scriptName, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Thread context not found error type
 *
 * Specialized for thread context not found scenarios
 * Inherited from NotFoundError for backward compatibility.
 */
export class ThreadContextNotFoundError extends NotFoundError {
  constructor(
    message: string,
    threadId: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "ThreadContext", threadId, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Checkpoint not found error type
 *
 * Specialized for checkpoint not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class CheckpointNotFoundError extends NotFoundError {
  constructor(
    message: string,
    checkpointId: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "Checkpoint", checkpointId, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Trigger template error type not found
 *
 * Specialized for trigger template not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class TriggerTemplateNotFoundError extends NotFoundError {
  constructor(
    message: string,
    templateName: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "TriggerTemplate", templateName, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

/**
 * Node template not found error type
 *
 * Specialized for node template not found scenarios
 * Inherited from NotFoundError to maintain backward compatibility
 */
export class NodeTemplateNotFoundError extends NotFoundError {
  constructor(
    message: string,
    templateName: string,
    context?: Record<string, unknown>,
    severity?: ErrorSeverity,
  ) {
    super(message, "NodeTemplate", templateName, context, severity);
  }

  protected override getDefaultSeverity(): ErrorSeverity {
    return "error";
  }
}

File: crates/wf-types/src/errors/resource_errors.rs
use crate::errors::WfError;

pub fn not_found_error(
    message: impl Into<String>,
    resource_type: String,
    resource_id: String,
) -> WfError {
    let mut ctx = std::collections::HashMap::new();
    ctx.insert(
        "resource_type".into(),
        serde_json::Value::String(resource_type),
    );
    ctx.insert("resource_id".into(), serde_json::Value::String(resource_id));
    WfError::new(message).with_context(ctx)
}

pub fn workflow_not_found_error(message: impl Into<String>, workflow_id: String) -> WfError {
    not_found_error(message, "Workflow".into(), workflow_id)
}

pub fn node_not_found_error(message: impl Into<String>, node_id: String) -> WfError {
    not_found_error(message, "Node".into(), node_id)
}

pub fn tool_not_found_error(message: impl Into<String>, tool_id: String) -> WfError {
    not_found_error(message, "Tool".into(), tool_id)
}

pub fn script_not_found_error(message: impl Into<String>, script_name: String) -> WfError {
    not_found_error(message, "Script".into(), script_name)
}

pub fn workflow_execution_not_found_error(
    message: impl Into<String>,
    execution_id: String,
) -> WfError {
    not_found_error(message, "WorkflowExecution".into(), execution_id)
}

pub fn checkpoint_not_found_error(message: impl Into<String>, checkpoint_id: String) -> WfError {
    not_found_error(message, "Checkpoint".into(), checkpoint_id)
}

pub fn trigger_template_not_found_error(
    message: impl Into<String>,
    template_name: String,
) -> WfError {
    not_found_error(message, "TriggerTemplate".into(), template_name)
}

pub fn node_template_not_found_error(message: impl Into<String>, template_name: String) -> WfError {
    not_found_error(message, "NodeTemplate".into(), template_name)
}

pub fn hook_template_not_found_error(message: impl Into<String>, template_name: String) -> WfError {
    not_found_error(message, "HookTemplate".into(), template_name)
}

pub fn agent_loop_not_found_error(message: impl Into<String>, agent_loop_id: String) -> WfError {
    not_found_error(message, "AgentLoop".into(), agent_loop_id)
}
